aspects/schema.go

```diff
@@ -0,0 +1,437 @@
+		}
+
+		v.choices = choices
+	}
+
+	return nil
+}
+
+type mapSchema struct {
+	baseSchema *CustomSchema
+
+	// entries map keys that can the map can contain to their expected types.
+	// Alternatively, the schema can instead key and/or value types.
+	entryTypes map[string]Schema
+
+	// valueType validates that the map's values match a certain type.
+	valueType Schema
+
+	// keyType validates that the map's key match a certain type.
+	keyType Schema
+
+	// requiredCombs holds combinations of keys that an instance of the map is
+	// allowed to have.
+	requiredCombs [][]string
+}
+
+func (v *mapSchema) Validate(raw []byte) error {
+	var level map[string]json.RawMessage
+	if err := json.Unmarshal(raw, &level); err != nil {
+		return err
+	}
+
+	var missing bool
+	for _, required := range v.requiredCombs {
+		missing = false
+		for _, key := range required {
+			if _, ok := level[key]; !ok {
+				missing = true
+				break
+			}
+		}
+
+		// met one combination of required keys so we can stop
+		if !missing {
+			break
+		}
+	}
+
+	if missing {
+		return fmt.Errorf(`cannot find required combinations of keys`)
+	}
+
+	if v.entryTypes != nil {
+		for key, val := range level {
+			if validator, ok := v.entryTypes[key]; ok {
+				if err := validator.Validate(val); err != nil {
+					return err
+				}
+			} else {
+				return fmt.Errorf(`unexpected field %q in map`, key)
+			}
+		}
+	}
+
+	if v.keyType != nil {
+		for k := range level {
+			rawKey, err := json.Marshal(k)
+			if err != nil {
+				return err
+			}
+
+			if err := v.keyType.Validate(rawKey); err != nil {
+				return err
+			}
+
+		}
+	}
+
+	if v.valueType != nil {
+		for _, val := range level {
+			if err := v.valueType.Validate(val); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (v *mapSchema) Parse(raw json.RawMessage) error {
+	var level map[string]json.RawMessage
+	if err := json.Unmarshal(raw, &level); err != nil {
+		return err
+	}
+
+	requiredRaw, ok := level["required"]
+	if ok {
+		if err := json.Unmarshal(requiredRaw, &v.requiredCombs); err != nil {
+			return fmt.Errorf(`cannot unmarshal map's "required" field: %v`, err)
+		}
+	}
+
+	// a map can have a "schema" constrain with specific values
+	if schema, ok := level["schema"]; ok {
+		var nextLevel map[string]json.RawMessage
+		if err := json.Unmarshal(schema, &nextLevel); err != nil {
+			return fmt.Errorf(`cannot unmarshal map's "schema" field: %v`, err)
+		}
+
+		v.entryTypes = make(map[string]Schema, len(nextLevel))
+		for key, value := range nextLevel {
+			validator, err := v.baseSchema.Parse(value)
+			if err != nil {
+				return fmt.Errorf(`cannot parse constraint for key %q: %w`, key, err)
+			}
+
+			v.entryTypes[key] = validator
+		}
+
+		return nil
+	}
+
+	// alternatively, it can constrain the type of its keys and/or values
+	keyDescription, ok := level["keys"]
+	if ok {
+		var err error
+		v.keyType, err = v.baseSchema.Parse(keyDescription)
```

Comment:
this needs to check that the keyType is a string type, also if type is omitted in keyDescription string should be assumed